Add notification area for JITMs on Sensei Home

fjorgemota opened the pull request:
Fixes #5674

### Changes proposed in this Pull Request

* Create notification area for JITMs on Sensei Home;
* Fix some issues with the notifications backend regarding broken URLs;
* Also create new notices component based on the `Sensei_Admin_Notices` code;

### Testing instructions

On a WP installation containing Sensei LMS based on this branch and Sensei Pro on the New Onboarding Wizard branch - with a license configured:

1. Visit Sensei Home;
2. Verify that no notices appear;
3. Checkout an older Sensei Pro version, let's say, 1.6.2;
4. Verify if a notice appears;
5. Verify what happens when you click on the "What's new" link (it should open the changelog page);
6. Disable Sensei Pro in the plugins list;
7. Verify if another notice appears;
8. Verify if, when clicking on the "Activate" link, it properly activates Sensei Pro;
9. Now, visit senseilms.com and disconnect the Sensei Pro license from there;
10. Verify if the notice changes once again;

Also verify if the styles are correct according to the design specified.

fjorgemota committed:
Add component to render notices

fjorgemota committed:
Refactor code to use event bubbling to catch events on all the sensei notices blocks

So we can dismiss even notices that appeared after the dom was ready to use.

fjorgemota committed:
Share nonce to dismiss notices with the frontend on Sensei Home

fjorgemota committed:
Enqueue script to dismiss notices automatically for us

fjorgemota committed:
Refactor code to use our own implementation of notices, and reorganize imports

fjorgemota committed:
Reorganize WordPress imports

fjorgemota committed:
Fix default type of the notices

Before it was an array, which is wrong, now it's an object.

fjorgemota committed:
Add margin-bottom to the notices section, and hide it if there are no notices

fjorgemota committed:
Remove margins of the notices on Sensei Home

fjorgemota committed:
Center content vertically on the notice

fjorgemota committed:
Include Sensei Circle Logo on the left of each notice on Sensei Home

fjorgemota committed:
Fix typo in JSDoc

fjorgemota committed:
Add Info Link to notices on Sensei Home

fjorgemota committed:
Consider primary as enabled by default

fjorgemota committed:
Open links on a new tab by default on Sensei Home

fjorgemota committed:
Remove decodeEntities from the Sensei Home Notices

fjorgemota committed:
Remove HTML encode from Notices API on Sensei Home

fjorgemota committed:
Fix activate plugin URL on Sensei Home Notices

fjorgemota committed:
Move notices stylesheets to another file

jom commented on assets/home/notices.js:
```diff
@@ -0,0 +1,153 @@
+	return (
+		<a
+			href={ action.url }
+			target={ action.target ?? '_blank' }
```
I think we should default this to `_self` for internal actions (or all and we can set target otherwise). For example, for activations and updates they should stay on the same page.

fjorgemota replied:
Hey @jom! Thanks for the review.

The initial implementation was setting this as `_self`, but then I changed to `_blank` because of [this comment](https://github.com/Automattic/sensei/issues/5949#issuecomment-1281195349)  made by @Luchadores, which mentioned that he prefers to open all the links in a new tab.

Naturally, I can change back if you still prefer to go that way, as I'm not sure if this also applies to notices. But I think it's worth mentioning. :)

jom replied:
I can see it for external links, but I think it is a pretty serious [anti-pattern/accessibility issue](https://www.w3.org/TR/WCAG20-TECHS/G200.html) for internal links because it forces me to open in a new tab vs giving me the power to do it myself with a CMD/Ctrl key etc. In this case, I think we definitely usually want to because the action will change the state of the notice so we want them to come back and see the notice cleared.

fjorgemota replied:
Okay, good points. Changed the default in bfab7e0f0d8855f3269c96bee4c6be175acee9b9 - Thanks!

jom commented on includes/admin/home/notices/class-sensei-home-notices.php:
```diff
@@ -240,10 +240,7 @@ private function get_deactivated_plugin_update_notice( array $plugin_data ) {
-				self_admin_url( 'plugin.php?action=activate&plugin=' . $plugin_file ),
-				'activate-plugin_' . $plugin_file
-			),
+			'url'   => add_query_arg( '_wpnonce', wp_create_nonce( 'activate-plugin_' . $plugin_file ), self_admin_url( 'plugins.php?action=activate&plugin=' . $plugin_file ) ),
```
We'll need this for the update link as well, if not decoding later.

fjorgemota replied:
Good catch! Fixed in [dbcc304](https://github.com/Automattic/sensei/pull/5991/commits/dbcc304225e2dbc608e1e920c535eaab94e20c67)

jom reviewed (changes requested):
This is looking great! I think we'll need to clear after the heading. I left a couple of other minor comments, but it is really close.

![Screen Shot 2022-10-21 at 6 08 00 PM](https://user-images.githubusercontent.com/68693/197250917-dd4154ad-7e68-4669-9cc8-fb455947d4e4.png)

fjorgemota committed:
Fix update URl to not be escaped as HTML

fjorgemota committed:
Adjust heading of the notice so the heading is on a separate line

fjorgemota commented:
Just as a FYI, I noticed that, before, the heading was showing up on the same line as the message.

I fixed this in a9fb318, and now it looks like this:

![image](https://user-images.githubusercontent.com/529864/197255928-f42f5044-7700-4c21-97b8-9ea9ef8db90e.png)

fjorgemota committed:
Open links via actions in the same tab by default in Sensei Home notices

fjorgemota committed:
Merge branch 'feature/new-onboarding-experience' into add/jitm-frontend

jom commented:
The header looks much better, but when it is multi-lined it feels a bit scrunched:
![Screen Shot 2022-10-21 at 7 16 44 PM](https://user-images.githubusercontent.com/68693/197262695-4832a675-85f2-4820-8796-9acd34ad9048.png)
vs what that looks like on courses page:
![Screen Shot 2022-10-21 at 7 14 54 PM](https://user-images.githubusercontent.com/68693/197262733-149849bc-d287-4dfd-a2fe-362425bb5fb8.png)

Additionally, I just realized we need to take into account the `level` when choosing the left-side border color (per the designs):
![image](https://user-images.githubusercontent.com/68693/197262988-0eb2f0ed-1fe4-4237-bea4-6da014a02d70.png)

fjorgemota committed:
Fix Notices component to consider levels of the notices on display

fjorgemota committed:
Fix notice level not appearing consistently

fjorgemota committed:
Fix notice SVG being on the wrong element and with the wrong size in Sensei Home Notices

fjorgemota committed:
Allow notices to be dismissible by default if we have Sensei_Admin_Notices

Because that's the traditional behavior.

fjorgemota committed:
Don't add actions container if we don't have any actions to show

fjorgemota committed:
Remove unneeded CSS class related to the logo in notices in Sensei Home

fjorgemota committed:
Add some padding to the notices shown in Sensei Home

So even the notices in Sensei Blocks are beautiful.

fjorgemota commented:
Hey @jom! Thanks for your review.

> The header looks much better, but when it is multi-lined it feels a bit scrunched:

I noticed that this happens because the notice on the courses page is always dismissible. Meanwhile, the notices on Sensei Home weren't. I fixed this by:

1. changing the default so we have dismissible notices by default on Sensei Home if `Sensei_Admin_Notices` exists - in [c3acd52](https://github.com/Automattic/sensei/pull/5991/commits/c3acd523b245e7d40fa962baf8b577cffc264b22); 
2. Add the padding by default, even on non-dismissible notices, so it doesn't look ugly on Sensei Blocks - [3149768](https://github.com/Automattic/sensei/pull/5991/commits/31497685d6c8244fc75a921b15e76f46ff68c8d7);

> Additionally, I just realized we need to take into account the level when choosing the left-side border color (per the designs):

Regarding this, I fixed this on [ff53ab9](https://github.com/Automattic/sensei/pull/5991/commits/ff53ab912917d50d42a831c03b4f88ae75c65357) and [240d3b6](https://github.com/Automattic/sensei/pull/5991/commits/240d3b648dd02f7a4b38aa50f5e50afa3709d8e9), by using the correct field names to show the level of the notice. It should be fine now. :)

jom reviewed (approved):
This looks great! You also fixed the icon issue, which I was just about to mention 💯

Luchadores commented:
This is awesome! 🙏
